### DISCUSSION TITLE
Module development guide

### DISCUSSION THREAD

**Discussion by daviesgeek** [kg4vdk/arcos-linux-modules#22] [General Discussion]
Hi there,

I'm wondering if you could provide some guidance on adding / developing modules. I see the template module, but am unclear about how it's best handled to install things like a different browser, or VSCode or something like that. For things like VSCode, it would be great to use the existing package installation so you can just update via the main release channel rather than downloading another .deb file.

I'd also like to install LinBPQ and am a little lost on how to best load that as a module in arcOS.

The in-built module work fantastically well, i love that i can just boot and be transmitting packets in an extremely short amount of time with little to no configuration. Thanks for your hard work on this, can't wait to help contribute!

**Comment by kg4vdk** [comment]
Hey Matthew,

I will work on providing better guidance/templating.

I am attaching a module for VSCode. The module DOES NOT include the deb package (as it is nearly 100MB). To use the module, extract the `.tar.gz` file, and place the VSCODE directory inside the USER module directory. Then, move the `00_VSCODE.sh` up one level, so that the script resides in the USER directory. If run without the deb package in the `packages` directory, the user will be prompted to download the deb file (link provided in notification) and retry.

Maintaining updates of VSCode via the repo supplied by the deb file will likely NOT work. My recommendation would be to either turn off checking for updates, or to manually download the new deb file if an update becomes available AND is necessary for functionality. 

This VSCode module should be persistent as far as settings in VSCode. The necessary directories are created inside the module directory, and symlinked into the correct locations in the $HOME directory.

As for LinBPQ, there's no reason it can't be handled by a module. I have built out a LinBPQ Winlink Gateway in the past (https://github.com/kg4vdk/winlink-rms-gateway). You may be able to read through what's happening in that script to put something together that works for your particular use case. 

At the most basic level, any module is just a way to automate the steps required to reproduce the desired result. Knowing which files/directories are involved with the desired result can be tricky sometimes. arcOS includes a script to help track this sort of thing down. You can run `watch-dir /path/to/watch` to see if any files/directories are created/modified in a particular path. This can be very helpful when trying to figure out where a particular application is writing its config information and such.

Hopefully, at least some of this is helpful! :-)

Let me know if there's more information I can provide.

[VSCODE.tar.gz](https://github.com/user-attachments/files/16681468/VSCODE.tar.gz)

**Reply by daviesgeek** [reply]
Awesome thanks, I was actually able to get LinBPQ running great on arcOS via [the modernham's guide](https://themodernham.com/install-linbpq-bbs-packet-node-on-debian-ubuntu-and-raspbian/), but don't 100% know how to get this working via a module configuration for a couple different reasons:

- I'm not sure how to handle dependencies needed from apt. I'm assuming I'm going to have to copy the dependencies into the module?
- I'm not entirely sure how to have direwolf automatically start (like when opening YAAC or anything like that)

**Reply by kg4vdk** [reply]
For dependencies, you can use `apt download pkgname(s)`. Then, like in the VSCODE module, you can do the install from the locally held packages. 

For auto-starting Direwolf, check out the `start-yaac` script in `/opt/arcOS/bin`. That might help point the way.

**Reply by daviesgeek** [reply]
Ah okay, will do that, thanks so much for the support, it's very much appreciated!
